README.md

```diff
@@ -1,13 +1,408 @@
+```
+
+The main abstraction introduced by this library is Collection. The capitalization om the word 'Collection' is important to distinguish it from collection. The Collection is different from standard collections in C++ in that:
+
+1. It is functional in nature. All operations on a Collection produce some form of result by modifying the source in memory. 
+2. It is concise. After introducing lambda functions in C++11, and functors long before that, C++ has emerged as a viable language in which to write programs that conform to the functional paradigm. There is no template overhead to calling functions.
+3. Laziness seeking. We use laziness to optimize many Collections operations. So calling slice on the first five consecutive numbers will not require all numbers in the list to be examined. 
+4. Potentially unbounded. While all collections have a finite size, Collections do not. Short circuiting and infinite Collection by calling <section>head</section> on an infinite Collection. 
+[comment]: <> (Hopefully we can get infinite lists to work well)
+
+### Fluent Interfaces
+[Martin Fowler][mf] coined the term **fluent inteface** as a semantic facade that allows you to apply multiple properties to an object without having to redefine the object each time. Typically they are used to build domain specific languages, fluent interfaces provide not only more discoverable and readable code - but an easy structure to use particularly for people who may not be familiar with working with functional ideas in C++. Developing C++ Collections will benefit teams that would like to use method chaining as a replacement for typical collection based operations with syntax and code that is easier to parse and uses method chaining to communicate the actions taking place sequentially - similar to a UNIX pipe.
+
+[mf]:http://martinfowler.com/bliki/FluentInterface.html
+
+This form of method cascading which originated in [Smalltalk][sm] allows a layer of syntactic sugar to be added to an otherwise repetitive piece of code. Cascasing prevents the need to list the object repeatedly as you can have numerous methods communicating exactly what they want to convey all acting on a single object by returning and propagating `this`.
+
+[sm]: http://en.wikipedia.org/wiki/Smalltalk
+
+
+## Design
+
+The C++ Collection class uses `std::vector<T>` to store data internally. This allows a number of different constructors to transfer information to the vector to control pointer management and data organization. Additionally, this allows lists, arrays, and C-styled arrays to be stored and mapped to the C++ Collection. 
+
+Within the Collection, there are three main types of operations which are elaborated below: source operators, intermediate operators, and terminal operators. Source operators allow the creation of Collections. Intermediate operators work off Collections and return Collections. Finally, Terminal operators return a non-Collection object or no object in return. 
```

Comment:
I was trying to think of some sort of unique word to look at them as three different things that represent a pipe structure. Would you prefer constructors or generators @natebrennand ?